sos/plugins/juju_controller.py

```diff
@@ -0,0 +1,157 @@
+# This file is part of the sos project: https://github.com/sosreport/sos
+# Copyright (C) 2019 Nick Niehoff <nick.niehoff@canonical.com>
+#
+# This copyrighted material is made available to anyone wishing to use,
+# modify, copy, or redistribute it subject to the terms and conditions of
+# version 2 of the GNU General Public License.
+#
+# See the LICENSE file in the source distribution for further information.
+
+import re
+import os
+from sos.plugins import Plugin, UbuntuPlugin
+
+
+# A Juju Controller is a juju machine so this plugin assumes the juju_machine
+# plugin will be executed to collect that information
+class JujuController(Plugin, UbuntuPlugin):
+    """ Juju orchestration tool - Controller
+    """
+
+    plugin_name = "juju_controller"
+    profiles = ("virt", "sysmgmt")
+
+    # Use files instead of packages here because there is no package installed
+    #  on a juju controller that uniquely identifies it:
+    files = "/var/lib/juju/db"
+
+    option_list = [
+        (
+            "export-mongodb",
+            "Export mongodb collections as json files",
+            "",
+            False,
+        )
+    ]
+
+    def export_mongodb(self, juju_machine_id, dbpass):
+        # We are gonna grab the output of the mongo command
+        # and do some regexp and replace()/split() it to a list.
+        # Usually when connecting with ssl on localhost, the cli send some
+        # output which unfortunately ends up in stdin instead of stderr
+        # and we need to do this extra work to parse it.
+        collections_query = ("mongo 127.0.0.1:37017/juju "
+                             + "--authenticationDatabase admin "
+                             + "--ssl --sslAllowInvalidCertificates --quiet "
+                             + "--username %s --password %s "
+                             % (juju_machine_id, dbpass)
+                             + "--eval 'db.getCollectionNames();'")
+        collections_output = os.popen(collections_query).read()
```

Comment:
Nack to `popen()` calls directly in the plugin. Use the existing `get_command_output()` in the `Plugin` class. E.G.

```python
_collections = self.get_command_output(collections_query)
if _collections['status'] == 0:
    collections_output = _collections['output']
else:
    collections_output = ''
```

Reply:
As a practical example to copy&paste&modify:

https://github.com/sosreport/sos/blob/master/sos/plugins/pulp.py#L72-L75
https://github.com/sosreport/sos/blob/master/sos/plugins/pulp.py#L111
https://github.com/sosreport/sos/blob/master/sos/plugins/pulp.py#L116-L121